File: c/ObjectC/QiSpaceObjectiveC/Sequr_Demo.m
//
//  Sequr_Demo.m
//  QiSpace_ObjectiveC_Demo
//
//  Created by Roy Zheng on 2022-06-12.
//

#import "Sequr_Demo.h"
#import "sequrObjCWrapper.h"
#import "ViewController.h"
#include "qispace_info.h"

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#define SUBKEY_BUFFER_SIZE 1300
#define QE_BUFFER_LEN 50
#define RND_LEN 50


@interface Sequr_Demo ()

@end

@implementation Sequr_Demo

/**
 * Decoding any payload from QiSpace RESTful API utilizes SEQUR decode().
 * This demo page demonstrates the process of decoding Qantum Entropy  (or any QiSpace payload) with SEQUR
 * Furthermore, the setup of the subkey and QE beginning from REST API to storage is also demonstrated.
 *
 * Prerequisite: An enterprise device token for authentication when interfacing QiSpace REST API
 *
 * Logical flow:
 * 1. Press 'Setup subkey' to establish and unpack the subkey with QiSpace Enterprise such that QE can be decrypted
 * 2. Press 'Get QE' to retrieve and unpack the Quantum Entropy from QiSpace Enterprise in order to later seed Pseudo-Quantum-Random-Number-Data (PQRND) API
 * 3. Press 'Get PQRND' to generate pseudo quantum random data
 */

@synthesize decoded_qe_text, random_string_text;

// Attributes
// App client and server shoud have the same token and subkey at all times
// Assume token is given for this demo app
// For real applications, the token and subkey are initialize during application setup
uint8_t *subkey = NULL;
size_t subkey_len;
uint8_t *qe;
int qe_len_val;
uint8_t *random_bytes = NULL;
pQrnd_Handle randHandle = NULL;

- (void)viewDidLoad {
    [super viewDidLoad];
    //Setup initializations
    qe = NULL;
    random_bytes = (uint8_t *)calloc(RND_LEN, 1);}

/**
 * Stores subkey from QiSpace Enterprise REST API using device bearer token
 *
 *  @param token the  bearer token for this specified Enterprise device used as authentication to obtain subkey from QiSpace REST API
 *  @param subkey the byte array  where the operable subkey will be stored
 *  @param subkey_len the length of the subkey's byte array
 */
void get_subkey(NSString *token, uint8_t *subkey, size_t *subkey_len){
    NSString *base = [NSString stringWithUTF8String: QISPACE_ENTERPRISE_BASE_URL];
    NSString *subkey_endpoint = [NSString stringWithFormat:@"%@/api/v1/sub_key", base];
    
    // QiSpace REST API Call to obtain subkey
    /**
     *       Sends a POST request to obtain
     *       the subkey, a necessary component to decrypting the payload,
     *       in hex string format.
     */
    NSMutableURLRequest *subkey_request = [[NSMutableURLRequest alloc] initWithURL: [NSURL URLWithString: subkey_endpoint]];
    
    // Configuring QiSpace subkey REST API request with specifications
    [subkey_request setHTTPMethod: @"POST"];
    [subkey_request setValue:@"application/json"
          forHTTPHeaderField:@"Content-Type"];
    [subkey_request setValue:token
          forHTTPHeaderField:@"Authorization"];
    
    NSURLSession *subkey_session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]
                                                          delegate:nil
                                                     delegateQueue:[NSOperationQueue mainQueue]];
    
    NSURLSessionDataTask *subkey_task = [subkey_session dataTaskWithRequest:subkey_request
                                    completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSHTTPURLResponse *subkey_response = (NSHTTPURLResponse *) response;

        // Verify the HTTP response from QiSpace Enterprise
        if((long)[subkey_response statusCode] != 200){
            NSLog(@"Error with URL RESPONSE");
            return;
        }
        
        // Processing and extracting json payload as C-string
        NSDictionary *subkey_payload = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
        NSString *subkey_string = [subkey_payload objectForKey: @"sub_key"];
        const char *subkey_c_string = [subkey_string UTF8String];

        *subkey_len = strlen(subkey_c_string)/2;
        
        // Convert subkey hex string to byte array
        convert_hex_to_bytes(subkey, strlen(subkey_c_string), subkey_c_string);
    }];
    // Trigger HTTP request to establish subkey from QiSpace Enterprise REST API
    [subkey_task resume];
}


/**
 * Handles 'setup subkey' button press event
 * Prerequisite: the enterprise devicei token must be valid for authentication when interfacing with QiSpace REST API
 */
- (IBAction)event_subkey_pressed:(id)sender {
    if(subkey == NULL){
        subkey = (uint8_t *)calloc(SUBKEY_BUFFER_SIZE, sizeof(uint8_t));
    }
    
    get_subkey(token, subkey, &subkey_len);
}

/**
 *  Unpacks the retrieved QE to a displayable hex stirng format
 *
 * - Prerequisite:
 *  - subkey: must be retrieved before QE is decoded
 * @param data the HTTP request json data from QiSpace Enterprise containing an encrypted Base64 QE and its respective IV
 * @returns a hex string representation of the decrypted QE
 */
char *unpack_qe(NSData *data){
    // Process json as dictionary and extract the Quantum Entropy and its IV
    NSDictionary *qe_payload = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
    NSString *encoded_qe_string = [qe_payload objectForKey: @"payload"];
    NSString *iv_string = [qe_payload objectForKey: @"iv"];
    
    // Convert to C-string because conversion functions are C-based
    const char *encoded_qe_c_string = [encoded_qe_string UTF8String];
    const char *iv_c_string = [iv_string UTF8String];
    
    // Converting from base64 to byte array format
    size_t iv_len_val;
    uint8_t *encrypted_qe_byte_array = base64_decode_to_bytes(encoded_qe_c_string, strlen(encoded_qe_c_string), &qe_len_val);
    uint8_t *iv_byte_array = base64_decode_to_bytes(iv_c_string, [iv_string length], &iv_len_val);
    
    // Perform SEQUR decoding on QE
    int ret = SEQUR_decode(subkey, subkey_len, iv_byte_array, encrypted_qe_byte_array, qe_len_val, qe);
    
    // Formatting qe result as hex string for output
    return convert_bytes_to_hex(qe, qe_len_val);
}


// Handles 'Get QE' button press event
/** Prerequisites:
 * - subkey:  must be initialized to decode the retrieved QE
 * - token: a proper enterprise device token for authentication when interfacing QiSpace QE REST API
 */
- (IBAction)event_qe_pressed:(id)sender {
    // Validate prerequisites before continuing
    if(subkey == NULL){
        self.decoded_qe_text.text = @"Need to initialize subkey before getting QE";
        return;
    }
    if(qe == NULL){
        qe = (uint8_t *)calloc(QE_BUFFER_LEN, sizeof(uint8_t));
    }
    
    NSString *base = [NSString stringWithUTF8String: QISPACE_ENTERPRISE_BASE_URL];
    NSString *qe_endpoint = [NSString stringWithFormat: @"%@/api/v1/qe/50", base];
    
    /**
     *       QiSpace API response to a GET request consists of
     *       a payload, a pre-encrypted Base64 Quantum Entropy (QE) value,
     *       and an iv, initial vector Base64-encoded, which diminishes data encryption repetition
     */
    // Configuring QiSpace Quantum Entropy REST API request with specifications
    NSMutableURLRequest *qe_request = [[NSMutableURLRequest alloc] initWithURL: [NSURL URLWithString: qe_endpoint]];
    [qe_request setHTTPMethod: @"GET"];
    [qe_request setValue:@"application/json"
          forHTTPHeaderField:@"Content-Type"];
    [qe_request setValue:token
          forHTTPHeaderField:@"Authorization"];
    
    NSURLSession *qe_session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]
                                              delegate:nil
                                        delegateQueue:[NSOperationQueue mainQueue]];
    
    NSURLSessionDataTask *qe_task = [qe_session dataTaskWithRequest:qe_request
                            completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSHTTPURLResponse *qe_response = (NSHTTPURLResponse *) response;
        
        // Verify the HTTP response from QiSpace Enterprise
        if((long)[qe_response statusCode] != 200){
            NSLog(@"Error with URL RESPONSE");
        }
        
        // Obtain and output QE from HTTP data
        char *output = unpack_qe(data);
        self.decoded_qe_text.text = [NSString stringWithUTF8String:output];
        
        // Reset PQRND handle
        randHandle = NULL;
    }];
    
    [qe_task resume];
}

/**
 * Returns a pseudo quantum random data string by using QE as seed
 *
 * @param seed a value used to seed PQRND API in order to generate pseudo random data
 * @param seed_len the length of the seed used
 * @param random_bytes the variable holding the PQRND API generated data
 * @param aHandle  an object associated with PQRND API used during setup
 * @returns a hex string representation of the pseudo quantum random data generated by PQRND API
 */
NSString* get_pqrnd(uint8_t *seed, int seed_len, uint8_t *random_bytes, pQrnd_Handle *aHandle){
    int ret;
    // For the first time, seed PQRND with QE on startup
    if(randHandle == NULL){
        ret = pQrnd_setup(seed, seed_len, aHandle);
    }
    
    // Generate PQRND bytes per button press
    ret = pQrnd_generate(*aHandle, random_bytes, RND_LEN);
    
    // Return random bytes as hex
    char *random_str = convert_bytes_to_hex(random_bytes, RND_LEN);
    if(ret != 1){
        return @"An error occurred during PQRND";
    }
    return [NSString stringWithUTF8String: random_str];
    
    // Note: pQrnd will be cleaned up and closed after leaving this page
}

// Handles 'Get PQRND' button press event
- (IBAction)event_pqrnd_pressed:(id)sender {
    //Check that QE has been retrieved so it can be used as seed
    if(qe == NULL){
        random_string_text.text = @"Need to get QE before generating random string";
        return;
    }
    // Gets and displays pseudo quantum random data to screen using QE and pQrnd API handle
    random_string_text.text = get_pqrnd(qe, qe_len_val, random_bytes, &randHandle);
}

// Handles cleanup activities as View exits
-(void)viewWillDisappear:(BOOL)animated
{
    // Cleaning up pQrnd random generation
    if(randHandle != NULL){
        int ret = pQrnd_close(randHandle);
    }
    
    // Freeing allocated memory
    free(qe);
    free(random_bytes);
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: c/Swift/QiSpaceiOSDemoApp/SequrDemo.swift
//
//  SequrDemo.swift
//  QEEPiOSDemoApp
//
//  Created by Roy Zheng on 2022-06-01.
//

import SwiftUI

/// To decode any payload from QiSpace RESTful API, we use SEQUR decode().
/// This demo shows  SEQUR decoding applied to QE (Quantum Entropy) 
/// Prerequisite: A device token for QiSpace API call

/// Note: Only one subkey and one active token per enterprise device

/// Logical flow:
///  1. Press 'Setup subkey' to establish the sub_key between client and QiSpace Enterprise
///  2. Press 'Get QE' to get (and decode) the Quantum Entropy (QE) from QiSpace Enterprise
///  3. Press 'Get PQRND' to get the pQrnd string (by seeding it with the retrieved QE)

// Returns the hex string equivalent of a byte array
extension Data{
    func hexEncodedString() -> String{
        return map { String(format: "%02hhx", $0) }.joined()
    }
}

struct SequrDemo: View {
    let BGColor : Color = Color(red: 0/255, green: 0/255, blue: 0/255, opacity: 77.5/100)
    
    // QiSpace Endpoints:
    let sub_key_endpoint = qispace_info.base_url + "/api/v1/sub_key"
    let qe_endpoint = qispace_info.base_url + "/api/v1/qe/50"    // Note: QE size is specified to be 50 bytes for this demo
    
    // Attributes related for API
    @Binding var sub_key : [UInt8]
    @State var qe : [UInt8]? = nil
    @State var decoded_qe : String = ""
    @State var myHandle : pQrnd_Handle? = nil
    
    // Note: Token is set as this for this demo. Normally, token is generated upon device setup
    let token = "Bearer "+qispace_info.access_token
    
    // Define attributes for UI display
    @State var decode_string : String = "*some QE value*"
    @State var random_generated_string : String = "*a randomly generated string*"
    @State var random_str : [UInt8]? = [UInt8].init(repeating: 0, count: 32)
    
    var sequr_wrapper : sequrSwiftWrapper? = sequrSwiftWrapper()
    
    
    /// Retrieves, unpacks and stores the subkey from QiSpace Enterprise using Enterprise device's token
    ///
    /// - Prerequisite: Device token for QiSpace API authorization
    /// - Parameters: None
    /// - Returns: None
    func get_subkey(){
        //Setup HTTP request with all relevant data
        var sub_key_request = URLRequest(url: URL(string: sub_key_endpoint)!)
        sub_key_request.httpMethod = "POST"
        sub_key_request.httpBody = try? JSONSerialization.data(withJSONObject: [], options: [])
        sub_key_request.addValue("application/json", forHTTPHeaderField: "Content-Type")        // Note: Content-Type must be of application/json
        sub_key_request.addValue(token, forHTTPHeaderField: "Authorization")
        
        // Sends a POST request to obtain the sub_key, a necessary component to decrypting the payload, in the form of a hex string
        let sub_key_operation = URLSession.shared.dataTask(with: sub_key_request, completionHandler: { data, response, error -> Void in
            if let error = error{
                print("Request error: ", error)
                return
            }
            
            // After successful QiSpace response
            guard let response = response as? HTTPURLResponse else { return }
            if response.statusCode == 200 {
                DispatchQueue.main.async{
                    do{
                        // Process as json as dictionary and converts sub_key from hex string (raw API payload format) to byte array
                        let sub_key_payload = try JSONSerialization.jsonObject(with: data!) as! Dictionary<String, String>
                        self.sub_key = SequrDemo.hex_string_to_bytes(
                            sub_key_payload["sub_key"]!
                        )!
                    } catch {
                        print("error retrieving sub key")
                    }
                }
            }
        })
        
        // Triggers sub_key retrieval from QiSpace REST API
        sub_key_operation.resume()
    }
    
    
    
    /// Retrieves, unpacks and stores the Quantum Entropy (QE) from QiSpace Enterprise using device token and subkey
    ///
    /// - Prerequisites:
    ///  - Enterprise device authorization token: Needed for QE retrieval
    ///  - Sub_key: Required for decoding the encrypted QE
    /// - Parameters: None
    /// - Returns: None
    func get_qe(){
        print("Clicked get_qe")
        let sequr_wrapper = sequrSwiftWrapper()
        
        // Initializes HTTP Request object and provide specifications
        var qe_request = URLRequest(url: URL(string: qe_endpoint)!)
        qe_request.httpMethod = "GET"
        qe_request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        qe_request.setValue(token, forHTTPHeaderField: "Authorization")
        
        // Sends a GET request to retrieve Quantum Entropy (QE) which is later used to seed PQRND API and its IV from QiSpace Enterprise
        let qe_operation = URLSession.shared.dataTask(with: qe_request, completionHandler: { data, response, error -> Void in
            // Checks for QiSpace API request error and display
            if let error = error{
                print("Request error: ", error)
                return
            }
            
            // After successful QiSpace response
            guard let response = response as? HTTPURLResponse else { return }
            if response.statusCode == 200{
                DispatchQueue.main.async{
                    do{
                        let qe_payload = try JSONSerialization.jsonObject(with: data!) as? [String: Any]
                        
                        // Converts QE and respective IV from Base64 to byte array
                        var encrypted_QE : [UInt8]? = SequrDemo.base64_to_bytes(
                            string: String(describing: qe_payload!["payload"]!)
                        )
                        var qe_iv : [UInt8] = SequrDemo.base64_to_bytes(
                            string: String(describing: qe_payload!["iv"]!)
                        )
                        // Performs SEQUR decoding on QE using sub_key as QE is still encrypted
                        self.qe = [UInt8].init(repeating: 0, count: encrypted_QE!.count)        // Note: decrypted and encrypted byte array must be same length for decoding to work. Both are 50 bytes in this example
                        var ret = sequr_wrapper.decode(
                            key: &(self.sub_key),
                            iv: &qe_iv,
                            encoded_message: &encrypted_QE!,
                            decoded_message: &self.qe!
                        )
                        
                        // Formats the string for output
                        self.decoded_qe = String(Data(self.qe!).hexEncodedString())
                        
                        // Resets the PQRND handle
                        self.myHandle = nil
                    }catch{
                        print("error with qe/iv")
                    }
                }
            }
            
        })
        // Trigger QE retrieval request from the QiSpace Enterprise REST  API
        qe_operation.resume()
    }
    
    /// 'Get PQRND' button press event
    func pQrnd_btn_pressed(){
        // Displays output of getting PQRND
        random_generated_string = get_pQrnd(random_seed: &qe!, pqrnd_handle: &myHandle, rand_data: &random_str!)
    }
    
    
    /// Prepares and outputs the pQrnd to screen
    ///
    /// - Prerequisites;
    ///  - Quantum Entropy: Required to be used as seed during pQrnd setup (if first time)
    /// - Parameters:
    ///  - random_seed: a byte array to be used as a seed value (For this demo, QE is passed)
    ///  - pqrnd_handle: a pqrnd handle object for setting up pQrnd API
    ///  - rand_data: the future pseudo random quantum data
    /// - Returns: hex-encoded representation of PQRND data
    func get_pQrnd(random_seed seed: inout [UInt8], pqrnd_handle handle: inout pQrnd_Handle?, rand_data pqrnd: inout [UInt8]) -> String {
        
        // Step 1: Check QE has been retrieved prior
        if(qe == nil){
            return "Need to retrieve QE before generating random string"
        }

        // Seeds pQrnd for the new QE to generate a random value
        var ret : QEEP_RET
        if(myHandle == nil){
            ret = (sequr_wrapper?.pQrnd_setup(seed: &qe!, handle: &myHandle))!
        }
        
        // Generates random number per button press and display
        ret = (sequr_wrapper?.pQrnd_generate(handle: myHandle!, random_number_to_be: &random_str!))!
        
        return String(Data(random_str!).hexEncodedString())
        
        // Note: Remember to clean up pqrnd handle. This is done on back button press
    }
    
    @Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>
    
    var btn_back : some View {
        Button(action:{
            // Closes the pQrnd handle if used
            if myHandle != nil{
                let ret = sequr_wrapper?.pQrnd_close_handle(handle: myHandle!)
            }
            self.presentationMode.wrappedValue.dismiss()
        }) {
            HStack{
                Image("back_button_image")
                    .aspectRatio(contentMode: .fit)
                    .foregroundColor(Color.white)
                Text("Back")
            }
        }
    }
    var body: some View {
        
        ZStack{
            BGColor.ignoresSafeArea()
            VStack{
                Spacer()
                
                Button(action: get_subkey){
                    Text("SETUP SUB_KEY")
                        .padding()
                        .background(Color.gray)
                        .foregroundColor(.white)
                }
                
                Spacer()
                
                Button(action: get_qe){
                    Text("GET QE")
                        .padding()
                        .background(Color.gray)
                        .foregroundColor(.white)
                }
                
                VStack{
                    Text("Decoded QE: ")
                    Text(decoded_qe)
                        .font(.caption)
                        .padding()
                }.foregroundColor(Color.white)
                
                
                Spacer()
                
                HStack{
                    Spacer()
                        
                    Button(action: pQrnd_btn_pressed){
                        Text("GET PQRND")
                            .padding()
                            .background(Color.gray)
                            .foregroundColor(.white)
                    }
                    
                    VStack{
                        Text("pQrnd:")
                        Text(random_generated_string)
                            .font(.caption)
                            .padding(.top)
                    }.foregroundColor(Color.white)
                    
                    Spacer()
                }
                Spacer()
            }
        }
            .navigationBarBackButtonHidden(true)
            .navigationBarItems(leading: btn_back)
    }
    
    
    /// Returns the byte array equivalent of a hex string
    static func hex_string_to_bytes(_ string: String) -> [UInt8]? {
        let length = string.count
        if length & 1 != 0 {
            return nil
        }
        var bytes = [UInt8]()
        bytes.reserveCapacity(length/2)
        var index = string.startIndex
        for _ in 0..<length/2 {
            let nextIndex = string.index(index, offsetBy: 2)
            if let b = UInt8(string[index..<nextIndex], radix: 16) {
                bytes.append(b)
            } else {
                return nil
            }
            index = nextIndex
        }
        return bytes
    }
    
    /// Returns the byte array equivalent of a base64 string
    static func base64_to_bytes(string: String) -> [UInt8]{
        /// Convert base64 -> nsdata -> byte array
        let in_nsdata = Data(base64Encoded: string, options: .ignoreUnknownCharacters)
        let byte_arr = [UInt8] (in_nsdata!)
        return byte_arr
    }
}

//struct SequrDemo_Previews: PreviewProvider {
//    static var previews: some View {
//        let requests = Requests()
//        SequrDemo().environmentObject(requests)
//    }
//}
